### PR TITLE
Homework week3

### PR DESCRIPTION
**Pull request by AllouhM**

### CODE REVIEW

**Comment by rohni** [comment]
Hey Allouh,
It is clear that you have not followed the flow of creating a new week's homework in a new branch  based on master. 
```
git co master
git co -b week3  # creates a new branch
```

**Comment by AllouhM** [comment]
Hello, yes you are right, I told you in class that i forgot to do it for the first week homework but I did it for week2 and 3.

**Comment by rohni** [comment]
Hi Allouh,
Yeah, sorry about that.  I had not looked through all the files, I see the week2 stuff is not in there.  And fixing that problem, while a good exercise in learning some deep git skills, is not easy. 😄

File: Week3/homework/1-step3.js [hunk 1]
```diff
@@ -2,6 +2,8 @@
 
 function foo(func) {
   // What to do here? 
+  //As the argument is a function so I will call the argument itself.
+  return func();
 }
 
 function bar() {
```
Comment (rohni) [inline Week3/homework/1-step3.js]: Excellent.  Returning the result of calling func is very important, as the foo function does not (and should not) know what func does.

File: Week3/homework/2-step3.js [hunk 2]
```diff
@@ -3,6 +3,31 @@
...
+      sayThree();
+      sayFive();
+    } else if (i % 3 === 0) {
+      sayThree();
+    } else if (i % 5 === 0) {
+      sayFive();
+    }
+
+  }
+  console.log(values); // just to print out the array when the programme run
```
Comment (rohni) [inline Week3/homework/2-step3.js]: This is nice. Most people have been creating the values array first, and then processing the indexes (I also did this) but your way means you only traverse the whole structure once and do all your processing in one go.  Nice.

File: Week3/homework/3-step3.js [hunk 3]
```diff
@@ -1,25 +1,43 @@
...
+  let string = "";
+  if (num <= 0) return string;
 
-console.log('while', repeatStringNumTimesWithDoWhile('abc', 3));
+  do { string += str; num--; }
+  while (num > 0);
+
+  return string;
+}
+console.log('do..while', repeatStringNumTimesWithDoWhile('abc', 3));
```
Comment (rohni) [inline Week3/homework/3-step3.js]: Nice work on this, all perfectly done. 
Interesting that you decided on the last to test that num was greater than 0 and not before.
I would be inclined for readability to put your two statements (string +=..., and num--) on different lines.
But that is a personal choice.
Nice work.

File: Week3/homework/4-step3.js [hunk 4]
```diff
@@ -1,2 +1,22 @@
...
+const MotorBike = function () {
+    this.wheels = 2;
+    this.engines = 1;
+    this.seats = 1;
+
+};
+
+const forSaleMotorBike = new MotorBike();
+
+console.log(forSaleMotorBike);
```
Comment (rohni) [inline Week3/homework/4-step3.js]: Nicely done.

File: Week3/homework/5-step3.js [hunk 5]
```diff
@@ -1,2 +1,19 @@
...
+            product *= arr[i][k];
+        }
+    }
+    // Only change code above this line
+    return product;
+}
+
+// Modify values below to test your code
+const result = multiplyAll([[1, 2], [3, 4], [5, 6, 7]]);
+console.log(result);
```
Comment (rohni) [inline Week3/homework/5-step3.js]: Good job.

File: Week3/homework/6-step3.js [hunk 6]
```diff
@@ -4,3 +4,13 @@ const arr2d = [[1, 2], [3, 4], [5, 6]];
...
+// //We use recursive function so it iterates through the array and check every element if
+//it is array so it calls back the same function but with the array element as an argument(as it is also an array)
+// and so on until the number of nested arrays is finished it return the element's value of the last array. 
+
+function printArrayItems(arr) {
+    for (let i = 0; i < arr.length; i++)
+        Array.isArray(arr[i]) ? printArrayItems(arr[i]) : console.log(arr[i]);
+}
+
+printArrayItems(arr3d);
```
Comment (rohni) [inline Week3/homework/6-step3.js]: Very nicely done.  You are the first of my reviews so far to use recursion. 👍

File: Week3/homework/step4-bonus.js [hunk 7]
```diff
@@ -4,7 +4,18 @@ const values = ['a', 'b', 'c', 'd', 'a', 'e', 'f', 'c'];
...
+    for (let i = 0; i < arr.length; i++) {
+        if (arr2.indexOf(arr[i]) === -1) [
+            arr2.push(arr[i])
+        ]
+    }
+    return arr2;
+}
 // Replace `yourFunction` with the name of the function you just created
-const uniqueValues = yourFunction(values);
+const uniqueValues = unique(values);
```
Comment (rohni) [inline Week3/homework/step4-bonus.js]: ✅ For a bonus, can you do this using reduce?

File: Week3/homework/step4-bonus.js [hunk 8]
```diff
@@ -4,7 +4,18 @@ const values = ['a', 'b', 'c', 'd', 'a', 'e', 'f', 'c'];
...
 
 console.log(uniqueValues);
```
Comment (rohni) [inline Week3/homework/step4-bonus.js]: ✅ And for a bonus, is it possible to use map,filter, or reduce to accomplish this?

File: Week3/homework/step4.js [hunk 9]
```diff
@@ -2,7 +2,17 @@
...
+}
+
 const addSix = createBase(6);
 
-addSix(10); // returns 16
-addSix(21); // returns 27
+const result1 = addSix(10);
+console.log(result1); // returns 16
+const result2 = addSix(21);
+console.log(result2);// returns 27
```
Comment (rohni) [inline Week3/homework/step4.js]: 👍 And I like that you created a named internal function, it makes it much easier to read. 😄

**Comment by rohni** [comment]
Hi Allouh,
Very nice work, well done. 
I have noticed that you lean towards using for loops to cycle through things.  Something to think about in the future is trying out a more functional approach, and see how it feels.   I, personally, have found that using functional programming has let me focus more on what I want to solve, and not get too bogged down in the details of exactly how.  
You definitely understand the principles, as your use of recursion and higher order functions demonstrated.
Again well done, very nice work. 👍

**Pull request closed without merging**